### PR TITLE
Merge database tool

### LINKED ISSUE #22: Database Merge
Should be nice to be able to automatically merge 2 or more password databases.

Use case: You have a KeePassX database on your PC and a database on your Phone, you "sync" them with Syncthing, Dropbox, whatever and you will lose changes on PC or Phone version so you need to keep both databases aside. With the Merge feature, you open both database on KeePassX and merge them so now you have all password in 1 file.

### PR DESCRIPTION
**Pull request by droidmonkey**
## Description
This PR institutes a manner in which to merge an unopened database to the currently open and selected database. Fixes #22.

## Motivation and Context
This is an oft-requested feature and has been coded and waiting in the KeePassX repository for several years.

## How Has This Been Tested?
- [x] Travis CI

## Types of changes

_New feature (non-breaking change which adds functionality)_

## Checklist:

- :white_check_mark: My code follows the code style of this project.
- :white_check_mark: My change requires a change to the documentation.
- :heavy_minus_sign: I have updated the documentation accordingly.
- :white_check_mark: I have read the **CONTRIBUTING** document.
- :x: I have added tests to cover my changes.
- :white_check_mark: All new and existing tests passed.

:neckbeard:

### CODE REVIEW

File: src/core/Group.cpp [hunk 1]
```diff
@@ -440,6 +459,18 @@ QList<Entry*> Group::entriesRecursive(bool includeHistoryItems) const
     return entryList;
 }
 
+Entry* Group::findEntry(const Uuid& uuid)
+{
+    Q_ASSERT(!uuid.isNull());
+    Q_FOREACH (Entry* entry, m_entries) {
```
Comment (TheZ3ro) [inline src/core/Group.cpp]: We need to convert Q_FOREACH to C++11 foreach. (Q_FOREACH will be deprecated soon)

File: src/core/Group.cpp [hunk 2]
```diff
@@ -490,6 +521,42 @@ QSet<Uuid> Group::customIconsRecursive() const
     return result;
 }
 
+void Group::merge(const Group* other)
+{
+    // merge entries
+    Q_FOREACH (Entry* entry, other->entries()) {
```
Comment (TheZ3ro) [inline src/core/Group.cpp]: Same

File: src/core/Group.cpp [hunk 3]
```diff
@@ -490,6 +521,42 @@ QSet<Uuid> Group::customIconsRecursive() const
...
+        // entries are searched by uuid
+        if (!findEntry(entry->uuid())) {
+            entry->clone(Entry::CloneNoFlags)->setGroup(this);
+        } else {
+            resolveConflict(this->findEntry(entry->uuid()), entry);
+        }
+    }
+
+    // merge groups (recursively)
+    Q_FOREACH (Group* group, other->children()) {
```
Comment (TheZ3ro) [inline src/core/Group.cpp]: Same

File: src/core/Group.cpp [hunk 4]
```diff
@@ -490,6 +521,42 @@ QSet<Uuid> Group::customIconsRecursive() const
...
+            group->setParent(this);
+        }
+    }
+
+    Q_EMIT modified();
+}
+
+Group* Group::findChildByName(const QString& name)
+{
+    Q_FOREACH (Group* group, m_children) {
```
Comment (TheZ3ro) [inline src/core/Group.cpp]: Same

File: src/gui/DatabaseTabWidget.cpp [hunk 5]
```diff
@@ -211,6 +236,114 @@ void DatabaseTabWidget::importKeePass1Database()
...
+{
+    QSet<QString> changedFiles;
+
+    changedFiles = m_changedFiles.subtract(m_expectedFileChanges);
+    m_changedFiles.clear();
+
+    if (changedFiles.isEmpty())
+        return;
+
+    Q_FOREACH (DatabaseManagerStruct dbStruct, m_dbList) {
```
Comment (TheZ3ro) [inline src/gui/DatabaseTabWidget.cpp]: Same

File: src/gui/SettingsWidgetGeneral.ui [hunk 6]
```diff
@@ -56,54 +56,80 @@
     </widget>
    </item>
    <item row="5" column="0">
+    <widget class="QLabel" name="label_2">
```
Comment (TheZ3ro) [inline src/gui/SettingsWidgetGeneral.ui]: The name of this widget is generic (`label_2`) should be something else with sense.

**Comment by TheZ3ro** [comment]
Proposed a fix for the bugs I've found.
IMO me we can merge.

**Comment by TheZ3ro** [comment]
Any news about this?

**Comment by droidmonkey** [comment]
I haven't actually compiled this and tried it out yet. Il also that today and merge it in. Been really focused on the new search functionality :wink:

**Comment by TheZ3ro** [comment]
Ok no problem :wink:

File: src/gui/DatabaseTabWidget.cpp [hunk 7]
```diff
@@ -271,7 +272,7 @@ void DatabaseTabWidget::checkReloadDatabases()
     if (changedFiles.isEmpty())
         return;
 
-    Q_FOREACH (DatabaseManagerStruct dbStruct, m_dbList) {
+    for (DatabaseManagerStruct dbStruct: asConst(m_dbList)) {
```
Comment (droidmonkey) [inline src/gui/DatabaseTabWidget.cpp]: I get a crash here when I try to autoreload a database.

**Comment by droidmonkey** [comment]
I got a crash on autoreload, see comment above

File: src/gui/DatabaseTabWidget.cpp (hunk 7 above)
Comment (TheZ3ro) [inline src/gui/DatabaseTabWidget.cpp]: Can you test if with the `Q_FOREACH` works or crash as well?

**Comment by TheZ3ro** [comment]
@droidmonkey for me crash with the following message: `'src/keepassx test2.kdbx' terminated by signal SIGSEGV (Address boundary error)`

Also you are right about the tainting.
A nice thing should be adding a button to merge database in the "discard my changes and load new database" dialog

File: src/gui/DatabaseTabWidget.cpp (hunk 7 above)
Comment (droidmonkey) [inline src/gui/DatabaseTabWidget.cpp]: The Q_FOREACH does not crash.

**Comment by droidmonkey** [comment]
I fixed all the issues noted, this is ready for a third set of eyes.

**Comment by TheZ3ro** [comment]
Do you have rebased the #67 in this PR? (I see commit related to Search)
If so we need to approve and merge that PR

**Comment by droidmonkey** [comment]
Yah I merged the search changes in and there were several conflicts. We should approve search first then this one.

**Comment by droidmonkey** [comment]
I intend to add the auto reload tests tonight and merge this in. Last call!

**Comment by droidmonkey** [comment]
I am getting crashes all over the place with the autoreload feature. This needs a lot more work before it can be merged.

**Comment by TheZ3ro** [comment]
Have you tried using *for* and removing the `asConst` from `m_dbList` ?

**Comment by droidmonkey** [comment]
Based on my debugging of the autoreload, these crashes occurred prior to the C++ 11 transition. The whole autoreload process needs to be redone in a different manner because it leaves behind a mess in the inner workings of the app.

I'm going to excise the auto reload part of this PR and move that code to its own feature branch. There is likely going to need a bit of replumbing in the way databases are loaded and metadata (in the tab widget) moved into proper classes. A lot of information gets out of sync by maintaining it in two different places.

**Comment by TheZ3ro** [comment]
Nice to know. (Ironically)

**Pull request merged**